### PR TITLE
Fix the NAMESPACE=1 release issue

### LINKED ISSUE #165: Lexical crash when building release with `NAMESPACE=1`
I found it in the project https://github.com/wingyplus/dagger_ex. When I try to formatting after opening an editor I found an error:

```
01:09:33.924 [error] ** (ErlangError) Erlang error: {:erpc, :noconnection}
    (kernel 8.5.3) erpc.erl:710: :erpc.call/5
    (server 0.1.0) lib/lexical/server/provider/handlers/formatting.ex:13: LXRelease.Server.Provider.Handlers.Formatting.handle/2
    (server 0.1.0) lib/lexical/server/provider/queue.ex:100: anonymous fn/2 in LXRelease.Server.Provider.Queue.State.as_task/2
    (elixir 1.14.3) lib/task/supervised.ex:89: Task.Supervised.invoke_mfa/2
    (elixir 1.14.3) lib/task/supervised.ex:34: Task.Supervised.reply/4
    (stdlib 4.2) proc_lib.erl:240: :proc_lib.init_p_do_apply/3

```

So, I open `.lexical/lexical.log` to see what's happens and found.

```
01:09:31.917 [error] Process #PID<0.2167.0> terminating
** (exit) exited in: GenServer.call(:"dagger_ex::node_process", {:start, ['/Users/thanabodee/src/github.com/lexical-lsp/lexical/_build/prod/rel/lexical/lib/common-0.1.0/ebin', '/Users/thanabodee/src/github.com/lexical-lsp/lexical/_build/prod/rel/lexical/lib/elixir_sense-2.0.0/ebin', '/Users/thanabodee/src/github.com/lexical-lsp/lexical/_build/prod/rel/lexical/lib/path_glob-0.2.0/ebin', '/Users/thanabodee/src/github.com/lexical-lsp/lexical/_build/prod/rel/lexical/lib/remote_control-0.1.0/ebin']}, 3500)
    ** (EXIT) an exception was raised:
        ** (FunctionClauseError) no function clause matching in IO.chardata_to_string/1
            (elixir 1.14.3) lib/io.ex:670: IO.chardata_to_string({:error, :bad_name})
            (elixir 1.14.3) lib/path.ex:548: Path.join/2
            (remote_control 0.1.0) lib/lexical/remote_control/project_node.ex:30: LXRelease.RemoteControl.ProjectNode.State.start/3
            (remote_control 0.1.0) lib/lexical/remote_control/project_node.ex:152: LXRelease.RemoteControl.ProjectNode.handle_call/3
            (stdlib 4.2) gen_server.erl:1149: :gen_server.try_handle_call/4
            (stdlib 4.2) gen_server.erl:1178: :gen_server.handle_msg/6
            (stdlib 4.2) proc_lib.erl:240: :proc_lib.init_p_do_apply/3
    (elixir 1.14.3) lib/gen_server.ex:1038: GenServer.call/3
    (remote_control 0.1.0) lib/lexical/remote_control/project_node.ex:106: LXRelease.RemoteControl.ProjectNode.start/3
    (remote_control 0.1.0) lib/lexical/remote_control.ex:24: LXRelease.RemoteControl.start_link/2
    (server 0.1.0) lib/lexical/server/project/node.ex:96: LXRelease.Server.Project.Node.start_node/1
    (server 0.1.0) lib/lexical/server/project/progress/support.ex:17: LXRelease.Server.Project.Progress.Support.with_progress/3
    (server 0.1.0) lib/lexical/server/project/node.ex:53: LXRelease.Server.Project.Node.init/1
    (stdlib 4.2) gen_server.erl:851: :gen_server.init_it/2
    (stdlib 4.2) gen_server.erl:814: :gen_server.init_it/6

```

### PR DESCRIPTION
**Pull request by scottming**
As we are using relative paths, we need to rewrite the boot file and two app-related definitions and references.

Fixes #165

### CODE REVIEW

File: apps/proto/lib/mix/tasks/namespace/release.ex [hunk 1]
```diff
@@ -42,7 +50,9 @@ defmodule Mix.Tasks.Namespace.Release do
     # matches if preceding characters involves either of: , " [ { [:blank:]
     # this way it doesn't match on substrings or directory names
     updated_contents =
-      String.replace(contents, ~r/([,"\[{[:blank:]])#{app_name}/, "\\1lx_#{app_name}")
+      contents
+      |> String.replace(~r/([,"\[{[:blank:]])#{app_name}/, "\\1lx_#{app_name}")
+      |> String.replace("Elixir.Lexical", "Elixir.LXRelease")
```
Comment (scohen) [inline apps/proto/lib/mix/tasks/namespace/release.ex:55]: Can we make this into a function and apply it to our dependencies as well?

File: apps/proto/lib/mix/tasks/namespace/release.ex [hunk 2]
```diff
@@ -2,15 +2,19 @@ defmodule Mix.Tasks.Namespace.Release do
...
   `remote_control.app` must be namespaced too
   """
+  alias Mix.Tasks.Namespace.Abstract
+
   use Mix.Task
-  @release_files ~w(start.script lexical.rel)
-  @apps_to_rewrite ~w(remote_control common server protocol proto)
+  @release_files ~w(start.script lexical.rel start_clean.script)
+  @boot_files ~w(start.boot start_clean.boot)
+  @apps_to_rewrite ~w(remote_control common server protocol proto lexical_shared lexical_plugin)
```
Comment (scohen) [inline apps/proto/lib/mix/tasks/namespace/release.ex:10]: this likely needs to be shared across modules.

**Review by scohen — COMMENTED** [review]
Thanks for tackling this. It looks good, but I think some of the code can be improved before we merge. As usual, my suggestions are below.

File: apps/proto/lib/mix/tasks/namespace/release.ex [hunk 3]
```diff
@@ -53,4 +63,40 @@ defmodule Mix.Tasks.Namespace.Release do
...
+        other ->
+          other
+      end)
+
+    updated_contents = :erlang.term_to_binary({script, script_info, new_module_infos})
+    File.write!(path, updated_contents)
+  end
+
+  def rewrite_deep_modules({:apply, {:application, mode, [{:application, app_name, app_info}]}}) do
+    new_modules = for module <- app_info[:modules], do: Abstract.rewrite_module(module)
```
[the remark this answers is not in the document]
Comment (scottming) [reply apps/proto/lib/mix/tasks/namespace/release.ex:94]: yes, this makes updating the value more explicit.

File: apps/proto/lib/mix/tasks/namespace/release.ex [hunk 4]
```diff
@@ -53,4 +63,40 @@ defmodule Mix.Tasks.Namespace.Release do
     namespaced_app_file = Path.join([ebin_path, "lx_" <> "#{app_name}.app"])
     :ok = File.rename(app_file_path, namespaced_app_file)
   end
+
+  defp update_boot_file(path) do
+    binary = path |> File.read!() |> :erlang.binary_to_term()
```
[the remark this answers is not in the document]
Comment (scottming) [reply apps/proto/lib/mix/tasks/namespace/release.ex:73]: nope, it is a binary

```elixir
iex(2)> "_build/dev/rel/lexical/releases/0.1.0/start_clean.boot" |> File.read!
<<131, 104, 3, 100, 0, 6, 115, 99, 114, 105, 112, 116, 104, 2, 107, 0, 7, 108,
  101, 120, 105, 99, 97, 108, 107, 0, 5, 48, 46, 49, 46, 48, 108, 0, 0, 0, 62,
  104, 2, 100, 0, 9, 112, 114, 101, 76, 111, 97, 100, 101, ...>>
  ```

[the remark this answers is not in the document]
Comment (scohen) [reply apps/proto/lib/mix/tasks/namespace/release.ex:73]: but you call `binary_to_term on it` thus, it's a term.

File: apps/proto/lib/mix/tasks/namespace/release.ex [hunk 5]
```diff
@@ -39,12 +47,19 @@ defmodule Mix.Tasks.Namespace.Release do
...
-    updated_contents =
-      String.replace(contents, ~r/([,"\[{[:blank:]])#{app_name}/, "\\1lx_#{app_name}")
-
-    File.write!(path, updated_contents)
+    contents
+    |> String.replace(~r/([,"\[{[:blank:]])#{app_name}/, "\\1lx_#{app_name}")
+    # Some deps app like `lexical_plugin` have `Elixir.Lexical` in their `app_file`
+    |> String.replace("Elixir.Lexical", "Elixir.LXRelease")
+    |> String.replace("Elixir.Sourceror", "Elixir.LXSourceror")
+    |> String.replace("Elixir.PathGlob", "Elixir.LXPathGlob")
```
Comment (scohen) [inline apps/proto/lib/mix/tasks/namespace/release.ex:62]: I think we need to think about this a little more.
Perhaps if we had a mapping from  app ->  root module we could make this a little more extensible.

File: apps/proto/lib/mix/tasks/namespace/task.ex [hunk 6]
```diff
@@ -0,0 +1,5 @@
+defmodule Mix.Tasks.Namespace.Task do
```
Comment (scohen) [inline apps/proto/lib/mix/tasks/namespace/task.ex:1]: put this in `Mix.Task.Namespace`

File: apps/proto/lib/mix/tasks/namespace/release.ex (hunk 4 above)
[the remark this answers is not in the document]
Comment (scottming) [reply apps/proto/lib/mix/tasks/namespace/release.ex]: Oic, naming issue. I misunderstood. fixed.

File: apps/proto/lib/mix/tasks/namespace/task.ex [hunk 7]
```diff
@@ -0,0 +1,5 @@
...
+  def apps_to_namespace do
```
Comment (scohen) [inline apps/proto/lib/mix/tasks/namespace/task.ex:2]: perhaps:
```elixir 
def app_mappings do 
  %{remote_control: Lexical.RemoteControl, common: Lexical.Common... sourceror: Sourceror}
end

def apps_to_namespace do 
  Map.keys(app_mappings())
end

def root_modules do 
   Map.values(app_mappings())
end

``` 
then your string replaces can be:

```elixir

Enum.reduce(...root_modules(), contents, fn root_module, contents -> 
  String.replace(contents, to_string(root_module), Namespace.Module.rewrite(root_module))
end)
```

File: apps/proto/lib/mix/tasks/namespace/module.ex [hunk 8]
```diff
@@ -0,0 +1,43 @@
+defmodule Mix.Tasks.Namespace.Module do
+  alias Mix.Tasks.Namespace
+
+  @apps_to_rewrite Namespace.Task.apps_to_namespace() -- ~w(proto protocol server)a
+
+  def rewrite(module_name) when module_name in @apps_to_rewrite do
```
Comment (scohen) [inline apps/proto/lib/mix/tasks/namespace/module.ex:6]: this should have been called `apply`. We're applying the namespace to the module.

File: apps/proto/lib/mix/tasks/namespace/module.ex [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+  def rewrite(module_name) do
+    module_name
+    |> Atom.to_string()
+    |> String.split(".")
+    |> maybe_rewrite_split_module(module_name)
+  end
+
+  @lexical "Lexical"
+
+  defp maybe_rewrite_split_module([@lexical], _) do
```
Comment (scohen) [inline apps/proto/lib/mix/tasks/namespace/module.ex:19]: having the module attribute doesn't help readability.

File: apps/proto/lib/mix/tasks/namespace/module.ex [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+    LXRelease
+  end
+
+  @deps_to_namespace ~w(Sourceror PathGlob)s
```
Comment (scohen) [inline apps/proto/lib/mix/tasks/namespace/module.ex:23]: move this to the top, and why the s after the end?

File: apps/proto/lib/mix/tasks/namespace/module.ex [hunk 11]
```diff
@@ -0,0 +1,43 @@
+defmodule Mix.Tasks.Namespace.Module do
+  alias Mix.Tasks.Namespace
+
+  @apps_to_rewrite Namespace.Task.apps_to_namespace() -- ~w(proto protocol server)a
```
[the remark this answers is not in the document]
Comment (scottming) [reply apps/proto/lib/mix/tasks/namespace/module.ex:4]: It is needed here, because `when module_name in @apps_to_rewrite` is guard.

**Review by scohen — APPROVED** [review]

**Pull request merged**